library/src/test/java/com/getstream/sdk/chat/models/UserComparisonTest.java

```diff
@@ -0,0 +1,78 @@
+package com.getstream.sdk.chat.models;
+
+import com.getstream.sdk.chat.rest.User;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.MockitoAnnotations;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class UserComparisonTest {
+
+    private static String TEST_USER_ID_1 = "testUserId_1";
+    private static String TEST_USER_ID_2 = "testUserId_2";
+    private static String TEST_USER_ID_3 = "testUserId_3";
+
+
+    private User x, y, z, x_clone;
+
+    @BeforeEach
+    void initTest() {
+        MockitoAnnotations.initMocks(this);
+
+        x = new User(TEST_USER_ID_1);
+        HashMap<String, Object> extraData = new HashMap<>();
+        extraData.put("group", "group chat");
+        x_clone = new User(TEST_USER_ID_1, extraData);
+        y = new User(TEST_USER_ID_2);
+        z = new User(TEST_USER_ID_3);
+    }
+
+
+    @Test
+    void equalsTest() {
+        assertTrue(x.equals(x));
+        assertTrue(x.equals(x_clone));
+        assertTrue(x_clone.equals(x));
+        assertTrue(x.equals(x_clone) == x_clone.equals(x));
+
+        assertFalse(x.equals(y));
+        assertTrue(x.equals(y) == y.equals(x));
+
+        y.setId(x.getId());
+        z.setId(y.getId());
+
+        if ((x.equals(y) && y.equals(z)))
+            assertTrue(x.equals(z));
+    }
+
+    @Test
+    void hashCodeTest() {
+        assertEquals(x.hashCode(), x_clone.hashCode());
```

Comment:
All tests are more or less about similar, so we might miss some line while supporting all of them. And we will have more tests. I would think about some generic test.
We can wrap objects into interface during tests:
```kotlin
interface EqualsHashObject {
  fun setId(String id)
  fun getId():String
  fun equals(Object o)
  fun hashCode(): Int
}
```
And define one test:
```kotlin
class TestEqualsHashUtil {
  fun test(a: EqualsHashObject, aClone: EqualsHashObject, b: EqualsHashObject){
    assertEquals(a.hashCode(), aClone.hashCode());
    //...
    a.setId(b.getId());
    assertEquals(a.hashCode(), b.hashCode());
  }
}
```
So all tests will look like this
```kotlin
class TestEqualsHashObjects {
  fun testChannels() {
    val channelA = wrapChannel(Channel());
    val channelAClone = wrapChannel(Channel());
    val channelB = wrapChannel(Channel());
    TestEqualsHashUtil.test(channelA, channelAClone, channelB);
  }

  fun wrapChannel(channel: Channel): EqualsHashObject {
    return object: EqualsHashObject {
      override fun setId (id: String) {
        channel.setId(id);
      }
    }
  }
}
```

Reply:
this is how you make unreadable tests; my suggestion is to write 1 complete set of tests for one type and then use ctrl-c and ctrl-v

Reply:
There is nothing to read about testing equals/hashCode tests. Java developers already for 25 years write such kind of utils, libraries and frameworks:
- https://jqno.nl/equalsverifier/
- https://gist.github.com/rherrmann/2970842
- and many more other private

Then Autovalue was developed by Google and now Kotlin is solving about the same problem internally.

Copy/paste will sooner or later trigger a bug.